### ISSUE TITLE
S108: False Positive when conditional compilation is used

### ISSUE THREAD

**Issue by rjgotten** [SonarSource/sonar-dotnet#4540]
### Description

"S108: Nested blocks of code should not be left empty" gives a false positive when conditional compilation is used with `#if` pragmas.

### Repro steps

Create anything using conditional compilation with nothing else within the method. E.g. something trivial like

```cs
public void LogInDebug(object obj)
{
#if DEBUG
    Trace.WriteLine( "Something to log only in debug", obj);
#endif
}
```

### Expected behavior

S108 is not raised, because clearly the block isn't (always) empty.

### Actual behavior

S108 is raised.

### Known workarounds

None. (Just mark it as a false positive and live with it, I suppose.)

### Related information

* C#/VB.NET Plugins version 8.13.1
* Visual Studio version N/A
* MSBuild / dotnet version ?
* SonarScanner for .NET version (if used) ?
* Operating System ?

(This is with SonarScanner running from a TeamCity build agent inside a Docker container. I have no information wrt those versions available to me.)

**Comment by costin-zaharia-sonarsource** [comment]
Hi @rjgotten, thanks for your feedback.

I've tried to add a test case for your issue here: https://github.com/SonarSource/sonar-dotnet/pull/4627 but I'm not able to reproduce the problem. It could be that I use a different version of `dotnet` and due to this a different compiler option. 

Can you please help us with the MsBuild/dotnet version? Maybe you can find it in the build logs.

**Comment by rjgotten** [comment]
@costin-zaharia-sonarsource 
I can maybe add a few `RUN` lines to the container file to dump version information that way.
I'll have a look.

**Comment by costin-zaharia-sonarsource** [comment]
Hi @rjgotten. I've managed to reproduce the issue this morning.
The conditional had to be in a nested block. See:

```
public void ConditionalCompilation(object obj)
{
   {
#if DEBUG
       Trace.WriteLine("message");
#endif
   }
}
```

**Comment by rjgotten** [comment]
Oops. I just looked back at the issue report and yeah; I posted the wrong reduced case.
The right one was using a local function - i.e. had a nested block.

That's my bad. Apologies.

(Well... at least we have a repro now.)

**Comment by costin-zaharia-sonarsource** [comment]
Yes, don't worry about that. Thanks for the feedback!
